src/analysis_and_optimization/Optimize.ml

```diff
@@ -1017,62 +1017,181 @@ let block_fixing mir =
+ * stmt: the MIR statement to optimize.
+ *)
+let optimize_mem_pattern
+    ~(gen_variables :
+          (int, Stmt.Located.Non_recursive.t) Map.Poly.t
+       -> int
+       -> string Set.Poly.t
+       -> string Set.Poly.t)
+    ~(update_stmt :
+          ( Expr.Typed.Meta.t Expr.Fixed.t
+          , (Expr.Typed.Meta.t, 'a) Stmt.Fixed.t )
+          Stmt.Fixed.Pattern.t
+       -> string Core_kernel.Set.Poly.t
+       -> ( Expr.Typed.Meta.t Expr.Fixed.t
+          , (Expr.Typed.Meta.t, 'a) Stmt.Fixed.t )
+          Stmt.Fixed.Pattern.t)
+    ~(update_expr : string Set.Poly.t -> Expr.Typed.t -> Expr.Typed.t)
+    ~(initial_variables : string Set.Poly.t) (stmt : Stmt.Located.t) =
+  let rev_flowgraph, flowgraph_to_mir =
+    Monotone_framework.inverse_flowgraph_of_stmt stmt
+  in
+  let fwd_flowgraph = Monotone_framework.reverse rev_flowgraph in
+  let (module Rev_Flowgraph) = rev_flowgraph in
+  let (module Fwd_Flowgraph) = fwd_flowgraph in
+  let soa_levels =
+    Monotone_framework.minimal_variables_mfp
+      (module Fwd_Flowgraph)
+      (module Rev_Flowgraph)
+      flowgraph_to_mir initial_variables gen_variables
+  in
+  let optimize_min_vars_stmt_base i stmt_pattern =
+    let variable_set = (Map.find_exn soa_levels i).exit in
+    let stmt_val =
+      Stmt.Fixed.Pattern.map (update_expr variable_set)
+        (fun x -> x)
+        stmt_pattern
+    in
+    update_stmt stmt_val variable_set
+  in
+  map_rec_stmt_loc_num flowgraph_to_mir optimize_min_vars_stmt_base
+    (Map.find_exn flowgraph_to_mir 1)
+
+(**
+  *)
+let gen_aos_variables
+    (flowgraph_to_mir : (int, Stmt.Located.Non_recursive.t) Map.Poly.t)
+    (l : int) (aos_variables : string Set.Poly.t) =
+  let mir_node mir_idx = Map.find_exn flowgraph_to_mir mir_idx in
+  let _, _, split_flowgraph = Map.split flowgraph_to_mir l in
+  match (mir_node l).pattern with
+  | Decl {decl_id; decl_type= Type.Sized _; _} -> (
+      let query_stmt stmt =
+        Mir_utils.query_aos_stmts decl_id aos_variables flowgraph_to_mir stmt
+      in
+      match Map.exists ~f:query_stmt split_flowgraph with
+      | true -> Set.Poly.singleton decl_id
+      | false -> Set.Poly.empty )
+  | _ -> Set.Poly.empty
+
+(**
+  * Deduces whether types can be Structures of Arrays (SoA/fast) or Arrays of Structs (AoS/slow)
+  *  See the docs in Mir_utils.query_aos_* functions for details on the rules 
+  *   surrounding when demotion from SoA -> AoS needs to happen.
+  * @param to_mir: A Map of flattened MIR statements whose 
+  *  key represents the flattened statements position in the nested structure
+  * @param 
+  *)
+let optimize_soa (mir : Program.Typed.t) =
+  let transform _ stmt =
+    optimize_mem_pattern ~gen_variables:gen_aos_variables
+      ~update_expr:Mir_utils.modify_soa_exprs
+      ~update_stmt:Mir_utils.mod_soa_stmt_pattern
+      ~initial_variables:Set.Poly.empty stmt
+  in
+  transform_program_blockwise mir transform
```

Comment:
`transform_program_blockwise` will apply the optimization to all blocks including UDFs, is that what you want? Otherwise, you can just apply `transform` to the logprob statement in the mir